File: src/2265.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

#include "tree.h"

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
   private:
    int answer = 0;

   public:
    pair<int, int> postorder(TreeNode* root) {
        pair<int, int> retval = {0, 0};
        if (root == nullptr) {
            return retval;
        }

        pair<int, int> left = postorder(root->left);
        pair<int, int> right = postorder(root->right);

        int subtreeSum = left.first + right.first + root->val;
        int subtreeNode = left.second + right.second + 1;
        if (root->val == subtreeSum / subtreeNode) {
            answer += 1;
        }
        retval = {subtreeSum, subtreeNode};

        return retval;
    }
    int averageOfSubtree(TreeNode* root) {
        int retVal = 0;

        if (root == nullptr) {
            return retVal;
        }

        answer = 0;
        postorder(root);
        retVal = answer;

        return retVal;
    }
};

int main(int argc, char** argv) {
    struct subject {
        vector<int> nums;
    };
    vector<subject> testData{{{4, 8, 5, 0, 1, numeric_limits<int>::min(), 6}}, {{1}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: root = [4,8,5,0,1,null,6]
     *  Output: 5
     *
     *  Input: root = [1]
     *  Output: 1
     */

    Solution cSolution;
    TreeNode* pRoot;
    int answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: root = [";
        for (long unsigned int j = 0; j < testData[i].nums.size(); ++j) {
            if (testData[i].nums[j] == numeric_limits<int>::min()) {
                cout << (j == 0 ? "" : ",") << "null";
            } else if (testData[i].nums[j] == numeric_limits<int>::max()) {
                continue;
            } else {
                cout << ((j == 0) ? "" : ",") << testData[i].nums[j];
            }
        }
        cout << "]\n";
        pRoot = buildBinaryTree(testData[i].nums, 0);

        answer = cSolution.averageOfSubtree(pRoot);
        cout << "Output: " << answer << "\n";

        cout << "\n";

        freeTreeNode(pRoot);
    }

    return EXIT_SUCCESS;
}

File: src/2265.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop

from tree import TreeNode


def logging_setting():
    LOG_LEVEL = logging.INFO  # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt="%Y/%m/%d %H:%M:%S",
    )


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self) -> None:
        self.answer = 0

    def postorder(self, root: Optional[TreeNode]) -> List[int]:
        if root == None:
            return [0, 0]

        leftSum, leftNode = self.postorder(root.left)
        rightSum, rightNode = self.postorder(root.right)

        subtreeSum = leftSum + rightSum + root.val
        subtreeNode = leftNode + rightNode + 1
        if root.val == subtreeSum // subtreeNode:
            self.answer += 1

        return [subtreeSum, subtreeNode]

    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        retVal = 0

        if root == None:
            return retVal

        self.answer = 0
        self.postorder(root)
        retVal = self.answer

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for root in [[4, 8, 5, 0, 1, None, 6], [1]]:
            # /* Example
            #  *  Input: root = [4,8,5,0,1,null,6]
            #  *  Output: 5
            #  *
            #  *  Input: root = [1]
            #  *  Output: 1
            #  */
            logging.info("Input: root = %s", root)

            retVal = pSolution.averageOfSubtree(TreeNode().buildTree(root))
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
